backend/controller/ingress.go

```diff
@@ -0,0 +1,67 @@
+package controller
+
+import (
+	"context"
+	"math/rand"
+	"strings"
+
+	"github.com/alecthomas/errors"
+
+	"github.com/TBD54566975/ftl/backend/common/slices"
+	"github.com/TBD54566975/ftl/backend/controller/dal"
+)
+
+func (s *Service) getIngressRoute(ctx context.Context, method string, path string) (*dal.IngressRoute, error) {
+	pathStart := strings.Split(path, "/")[1]
+	routes, err := s.dal.GetIngressRoutes(ctx, method, "/"+pathStart)
+	if err != nil {
+		return nil, err
+	}
+	matchedRoutes := slices.Filter(routes, func(route dal.IngressRoute) bool {
+		return matchURL(route.Path, path)
+	})
+	if len(matchedRoutes) == 0 {
+		return nil, dal.ErrNotFound
+	}
+
+	// TODO: add load balancing at some point
+	route := matchedRoutes[rand.Intn(len(matchedRoutes))] //nolint:gosec
+	return &route, nil
+}
+
+func getPathParams(pattern, urlPath string) (map[string]string, error) {
+	formatSegments := strings.Split(strings.Trim(pattern, "/"), "/")
+	urlSegments := strings.Split(strings.Trim(urlPath, "/"), "/")
+
+	if len(formatSegments) != len(urlSegments) {
+		return nil, errors.New("number of segments in pattern and URL path don't match")
+	}
+
+	params := make(map[string]string)
+	for i, segment := range formatSegments {
+		if strings.HasPrefix(segment, "{") && strings.HasSuffix(segment, "}") {
+			key := strings.Trim(segment, "{}")
+			params[key] = urlSegments[i]
+		}
+	}
+	return params, nil
+}
+
+func matchURL(pattern, urlPath string) bool {
```

Comment:
It would be good to not have two functions doing almost the exact same thing. Perhaps just have one that takes a closure with any matching path variables, eg.

```
func matchURL(pattern, urlPath string, matched func(key, value)) bool
```